### PR TITLE
[TT-Train] BERT for TTML: Initial commit. Compiles. Not tested.

### PR DESCRIPTION
**Pull request by epam-iaroslav-voitovych**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/26739

### Problem description
Implement a BERT model in TTML based on existing model implementations

### What's changed
Implementation is based on other examples in the TTML framework.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/models/bert.cpp [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    // Apply layer norm and dropout
+    embeddings = (*m_embedding_norm)(embeddings);
+    embeddings = (*m_embedding_dropout)(embeddings);
+
+    return embeddings;
+}
+
+autograd::TensorPtr Bert::process_attention_mask(const autograd::TensorPtr& attention_mask) const {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/models/bert.cpp:180]: this one could be slow, to mhy understanding for inference we don't use mask right?
If yes feel free to ignore it for now.

File: tt-train/sources/ttml/models/bert.cpp [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        auto num_heads = hidden_shape[1];
+        auto embedding_dim = hidden_shape[3];
+
+        // Slice to get only the [CLS] token: [batch, 1, 1, embedding_dim]
+        // Based on ttnn_fixed/distributed/ttnn_ops.cpp usage pattern
+        ttnn::SmallVector<uint32_t> start_indices = {0, 0, 0, 0};
+        ttnn::SmallVector<uint32_t> end_indices = {batch_size, num_heads, 1, embedding_dim};
+        ttnn::SmallVector<uint32_t> stride = {1, 1, 1, 1};
+
+        auto cls_token = ttnn::slice(hidden_states->get_value(), start_indices, end_indices, stride);
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/models/bert.cpp:249]: overall I dont trust ttnn::slice and ttnn::repeat function, if you can wirte some tests for our shapes to confirm they are working as expected it would be nice.

Comment (aliaksei-sala) [reply tt-train/sources/ttml/models/bert.cpp:249]: [Here](https://github.com/tenstorrent/tt-metal/tree/main/tests/ttnn/unit_tests/operations) is the folder with unit tests for different ops including slice and repeat. You can check if your shapes are included in these tests

File: tt-train/sources/examples/bert_example/main.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+using namespace ttml;
+
+int main() {
+    // Initialize device context
+    autograd::ctx().open_device();
+
+    // Create BERT configuration
+    models::bert::BertConfig config;
+    config.vocab_size = 30522;
+    config.max_sequence_length = 128;  // Smaller for example
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/bert_example/main.cpp:18]: default bert should fit without any issues, and I believe  batch_size 64 should work too

File: tt-train/sources/ttml/models/bert.cpp (hunk 2 above)
Comment (epam-iaroslav-voitovych) [reply tt-train/sources/ttml/models/bert.cpp:249]: Added the tests in the recent commit named "test: Add comprehensive tests for ttnn::slice and ttnn::repeat operat…"

File: tt-train/sources/examples/bert_example/main.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    config.num_blocks = 6;  // Smaller for example
+    config.dropout_prob = 0.1F;
+    config.use_token_type_embeddings = true;
+    config.use_pooler = false;  // Set true for classification tasks
+
+    // Create BERT model
+    auto bert_model = models::bert::create(config);
+
+    // Example input (batch_size=1, sequence_length=128 to match config)
+    std::vector<uint32_t> input_ids_data(128, 0);  // Initialize with padding
```
Comment (aliaksei-sala) [inline tt-train/sources/examples/bert_example/main.cpp:31]: Let's use a tokenizer instead of hardcoded numbers with comments.

File: tt-train/sources/ttml/models/bert.cpp [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        .embedding_dim = embedding_dim,
+        .sequence_length = max_sequence_length,
+        .dropout_prob = 0.0F,  // No dropout on position embeddings in BERT
+        .use_dropout_seed_per_device = true};
+    m_position_embeddings = std::make_shared<modules::TrainablePositionalEmbedding>(pos_config);
+
+    // Create token type embeddings if enabled
+    if (config.use_token_type_embeddings) {
+        // Type vocab size is typically 2 (sentence A and sentence B)
+        uint32_t type_vocab_aligned = ((config.type_vocab_size + 31) / 32) * 32;
```
Comment (aliaksei-sala) [inline tt-train/sources/ttml/models/bert.cpp:90]: It’s better to use named constants instead of hardcoded values
for example, use TILE_HEIGHT instead.

File: tt-train/sources/ttml/models/bert.cpp [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+                int layer_idx = std::stoi(layer_idx_str);
+
+                std::string layer_suffix = info.name.substr(layer_end + 1);
+
+                // Attention weights - BERT has separate Q, K, V, we need to combine them
+                if (layer_suffix == "attention.self.query.weight" ||
+                    layer_suffix == "attention.self.key.weight" ||
+                    layer_suffix == "attention.self.value.weight") {
+                    // Note: We'll need to handle QKV combination in a separate pass
+                    // For now, store them temporarily
```
Comment (aliaksei-sala) [inline tt-train/sources/ttml/models/bert.cpp]: What is the plan to implement it?

File: tt-train/sources/examples/bert_example/main.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+#include "models/bert.hpp"
+
+using namespace ttml;
+
+int main() {
+    // Initialize device context
+    autograd::ctx().open_device();
+
+    // Create BERT configuration
+    models::bert::BertConfig config;
```
Comment (aliaksei-sala) [inline tt-train/sources/examples/bert_example/main.cpp:16]: Let's use values from actual model configuration files.

File: tt-train/sources/ttml/models/bert.cpp [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+    const autograd::TensorPtr& input_ids, const autograd::TensorPtr& token_type_ids) {
+    // Validate input shapes
+    auto input_shape = input_ids->get_shape();
+
+    // Token embeddings
+    auto embeddings = (*m_token_embeddings)(input_ids);
+
+    // Add positional embeddings using the operator() which adds positions and applies dropout (no-op since
+    // dropout_prob=0)
+    embeddings = (*m_position_embeddings)(embeddings);
```
Comment (aliaksei-sala) [inline tt-train/sources/ttml/models/bert.cpp:156]: I believe `position_embeddings` should take `position_ids`, not the output of `token_embeddings`.
https://github.com/tenstorrent/tt-metal/blob/e96a6c2ce0c966e40d3ca63b710c129a41e699e8/models/demos/bert/demo/demo.py#L35

File: tt-train/tests/ops/slice_repeat_ops_test.cpp [hunk 9]
```diff
@@ -0,0 +1,603 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
Comment (aliaksei-sala) [inline tt-train/tests/ops/slice_repeat_ops_test.cpp:1]: ttnn ops should be tested in ttnn unit tests

File: tt-train/tests/ops/binary_ops_test.cpp [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+    // Test basic addition of two tensors
+    std::vector<float> data_a = {1.0f, 2.0f, 3.0f, 4.0f};
+    std::vector<float> data_b = {5.0f, 6.0f, 7.0f, 8.0f};
+
+    auto a = autograd::create_tensor(
+        core::from_vector(data_a, ttnn::Shape{1, 1, 2, 2}, device));
+    auto b = autograd::create_tensor(
+        core::from_vector(data_b, ttnn::Shape{1, 1, 2, 2}, device));
+
+    auto result = add(a, b);
```
Comment (aliaksei-sala) [inline tt-train/tests/ops/binary_ops_test.cpp:47]: This type of test verifies underlying ttnn ops, specifically ttnn::add here
I believe they duplicate the existing uint tests we have for ttnn ops

File: tt-train/tests/ops/gelu_op_test.cpp [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+    ASSERT_EQ(result_data.size(), expected.size());
+    for (size_t i = 0; i < expected.size(); ++i) {
+        EXPECT_NEAR(result_data[i], expected[i], 3e-2f);
+    }
+}
+
+// Block size alignment tests - testing Wt % 4 patterns
+// where Wt = ceil(C / 32) is the number of tiles in width dimension
+
+TEST_F(GELUOpTest, GELU_Aligned128) {
```
Comment (aliaksei-sala) [inline tt-train/tests/ops/gelu_op_test.cpp:190]: I think you can combine all tests which call `CompareGELUVsReferenceWithShape`  writing value parametrized [tests](https://google.github.io/googletest/advanced.html#how-to-write-value-parameterized-tests)

**Review by aliaksei-sala — CHANGES_REQUESTED** [review]

File: tt-train/tests/ops/binary_ops_test.cpp (hunk 10 above)
Comment (epam-iaroslav-voitovych) [reply tt-train/tests/ops/binary_ops_test.cpp:47]: We have wrappers over TTNN binary operations in TTML, so some testing should be good.
.

**Review by athompsonTT — COMMENTED** [review]
nanobind changes look good to me, thanks!

File: tt-train/sources/ttml/ops/multi_head_utils.cpp [hunk 12]
```diff
@@ -86,41 +86,52 @@ std::tuple<autograd::TensorPtr, autograd::TensorPtr, autograd::TensorPtr> heads_
         ttnn::SmallVector<uint32_t>{batch_size, 1, seq_len, embedding_dim * 3},
         ttnn::SmallVector<uint32_t>{1, 1, 1, 1});
 
-    // Reshape: (B, 1, S, E) -> (B, 1, S*H, E/H) -> (B, H, S, E/H)
-    // This works around the ttnn bug by doing reshape instead of nlp_create_qkv_heads
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/multi_head_utils.cpp]: what is the bug?
